### DISCUSSION TITLE
Non-continous GeoDataModules/GeoDatasets

### DISCUSSION THREAD

**Discussion by MathiasBaumgartinger** [torchgeo/torchgeo#2292] [Q&A]
I am trying to set up a `GeoDataModule` where the images/labels are _scattered_ and are not _spatially continous_.

I defined `__get_item__` using rasterio by querying the individual images manually depending on the query `BoundingBox`. However, when doing a test run, the queries go beyond the the individual tiles and I end up getting 

```
warnings.warn('shapes are outside bounds of raster. '
               'Are they in different coordinate reference systems?')
ValueError: Input shapes do not overlap raster.
```
![image](https://github.com/user-attachments/assets/1b9df45f-91df-47f5-b3b8-f07ea748f9fc)

Am I correct  with the conclusion that GeoDataModules need to be continous? So in case I have a patched dataset I would define it as `NonGeoDataset` although the individual images do have  _spatial_ information?

**Comment by adamjstewart** [comment]
> I defined `__getitem__` using rasterio by querying the individual images manually depending on the query `BoundingBox`.

You should not need to define your own `__getitem__`, the `RasterDataset.__getitem__` should handle everything you need. There is likely a bug in your `__getitem__`.

> the queries go beyond the the individual tiles

This is actually expected behavior for `GridGeoSampler`. We've gone back and forth on this a couple of times. In the original implementation, we simply skipped patches that didn't fit an entire patch inside. In #630, we changed this to get full tile coverage. #1329 changed this again so that we actually sample outside the bounding box of the tile. This was done so that we get full coverage of the tile, but don't overlap and compute metrics twice.

> Am I correct with the conclusion that GeoDataModules need to be continous? So in case I have a patched dataset I would define it as NonGeoDataset although the individual images do have spatial information?

GeoDatasets do not need to be continuous. You can use either GeoDataset or NonGeoDataset depending on whether you need to combine your dataset with another and depending on how large the patches are.

**Reply by MathiasBaumgartinger** [reply]
Thank you so much for your timely answer, @adamjstewart!

Definitely, `RasterDataset.__get_item__()` does return the image as expected. However, as I am working on a `SemanticSegmentationTask`, I do need an *image* as well as a *mask*. The [documentation](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.RasterDataset.is_image) clearly states:

> For datasets with both model inputs and outputs, a custom [__getitem__()](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#torchgeo.datasets.RasterDataset.__getitem__) method must be implemented.

Am I overthinking things? I would gladly accept a simpler solution to the problem :smile:. 

After writing the previous post I closely followed the implementation of the existing `Chesapeake Land Cove` dataset/module which seems to prevent the error, by taking into account the dataset projection. They first [increase the `patch_size`](https://github.com/microsoft/torchgeo/blob/94960bbcf6dd5e2008452144312c599db3b803b3/torchgeo/datamodules/chesapeake.py#L96) and in turn [crop each sample back to the original `patch_size` ](https://github.com/microsoft/torchgeo/blob/94960bbcf6dd5e2008452144312c599db3b803b3/torchgeo/datamodules/chesapeake.py#L94).

**Reply by adamjstewart** [reply]
Ah, we should update those docs. Can you open an issue for this (or even a PR to fix this)?

The current recommended strategy is to use 2 `RasterDatasets` (one for the image and one for the mask) and use `IntersectionDataset` to combine them. See L7 Irish, L8 Biome, and I/O Bench for examples of this in `torchgeo/datasets`.

**Reply by MathiasBaumgartinger** [reply]
FYI: https://github.com/microsoft/torchgeo/pull/2293

**Comment by Harshith-Reddy11** [comment]
Hi @MathiasBaumgartinger — this is a thoughtful question and still relevant!  

Here’s a cleaner way to handle geospatial image + mask data without overriding `__getitem__`:

1. **Use two separate `RasterDataset`s**, one for images and one for masks.  
2. **Combine them with `IntersectionDataset`**:
   ```python
   from torchgeo.datasets import RasterDataset, IntersectionDataset

   img_ds = RasterDataset(path="...", is_image=True)
   mask_ds = RasterDataset(path="...", is_image=False)
   ds = IntersectionDataset(img_ds, mask_ds)
3. **Let TorchGeo handle sampling** — use GridGeoSampler or RandomGeoSampler to fetch spatially aligned patches, even if they cross tile boundaries.

4. **Optional tweak**: If you prefer non-continuous behavior, crop or adjust using the patch_size parameter as seen in examples like L7Irish or I/O Bench.

This approach avoids custom __getitem__ logic and works smoothly — and it aligns with best practices endorsed by the maintainers.

🔗 Helpful references:

TorchGeo Docs: https://torchgeo.readthedocs.io/en/stable/api/generated/torchgeo.datasets.IntersectionDataset.html

Example (L7Irish): [https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/landsat.py](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/landsat.py?utm_source=chatgpt.com)

Hope this helps anyone working with fragmented spatial tiles! 🚀
